src/lib.rs

```diff
@@ -0,0 +1,386 @@
+extern crate libc;
+
+mod ffi;
+
+use std::path::Path;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+pub type Ts = ffi::rill_ts_t;
+pub type Key = ffi::rill_key_t;
+pub type Val = ffi::rill_val_t;
+pub type KV = ffi::rill_kv;
+
+// This function is kind-of insane. I really hope there's an easier
+// way to do this...
+fn err<T>() -> Result<T> {
+    let err = unsafe{ ffi::rill_errno_thread() };
+
+    let mut buf: [u8; 1024] = [0; 1024];
```

Comment:
Can't you just do something like:
```rust
let cstring = CString::new(vec![0; 1024]);
let clen = cstring.as_bytes().len();
let ptr = cstring.into_raw();
let len = unsafe {
    let len = ffi:rill_strerror(err, ptr, clen);
    if len >= clen { clen } else { len + 1 }
};
let cstring = unsafe { CString::from_raw(ptr) };

match cstring.to_str() {
    Err(err) => Err(format!("invalid error string: {}", err)),
    Ok(msg) => Err(String::from(msg)),
}
```
I don't know what's the best practise for checking strings coming from C:
Should we instead assume that the C library will produce a valid `UTF-8` string and panic if it doesn't?

Reply:
Note that performance doesn't matter too much in this case so the following is mostly academic. The issue I have with `CString::new(vec![0; 1024])` is that it ends doing two allocation a full copy of a bunch of 0 bytes. It bothers me a bit how much copying and allocating is required especially since the interface for `rill_strerror` is designed to minimise allocation and copying. In any case, not much we can do about it except avoid string manipulation in performance critical code paths.

As for whether the C code will produce valid UTF-8, that's really a case by case thing. In this case, I think it's actually possible for non-ascii to sneak its way into the string so I'd rather err on the side of caution.